design/ceph/update-osds-in-parallel.md

```diff
@@ -0,0 +1,174 @@
+# Updating OSDs in parallel
+**Targeted for v1.6**
+
+## Background
+In clusters with large numbers of OSDs, it can take a very long time to update all of the OSDs. This
+occurs on updates of Rook and Ceph both for major as well as the most minor updates. To better
+support large clusters, Rook should be able to update (and upgrade) multiple OSDs in parallel.
+
+
+## High-level requirements
+
+### Different update strategies
+Users will likely wish to specify different upgrade strategies. For many clusters, updating all OSDs
+on a node may be sufficient. For portable OSDs on PVCs, an OSD may not be associated directly with a
+node.
+
+Users may wish to update all OSDs in a particular failure domain at once, updating all OSDs in a
+domain before moving to the next domain. Some examples of failure domains are:
+ - by OSD (equivalent to the existing strategy)
+ - by node/host
+ - by rack
+ - by data center room
+ - by availability zone
+ - and many others
+
+> **Open question:** Will it be useful for users to try to update OSDs **across** failure domains?
+> I.e., update some number of OSDs in each failure domain instead of updating all OSDs within a
+> given failure domain?
+
+> **Open question:** Is it possible to identify a number of OSDs which can be safely updated in
+> parallel without regard to failure domain? For example, try to update some number (or percentage)
+> of OSDs at once in any failure domain. This might be a good default strategy ("best-effort"?).
+
+### Ability to set a maximumum number of OSDs to be updated simultaneously
+In the worst (but unlikely) case, all OSDs which are updated for a given parallel update operation
+might fail to come back online after they are updated. Users may wish to limit the number of OSDs
+updated in parallel in order to avoid too many OSDs failing in this way.
+
+### Cluster growth takes precedence over updates
+Adding new OSDs to a cluster should occur before updating currently-running OSDs. This allows users
```

Comment:
Instead of adding new OSDs before upgrades start, can't we add OSDs in parallel to the OSD updates? During the reconcile we know immediately if we need to update OSDs (on PVCs), but new OSDs have to first be prepared. As long as ok-to-stop says some OSDs can be stopped, seems like we can start the upgrade immediately of at least the first failure domain. When the new OSDs are done preparing, they could be started immediately. The next failure domain might be impacted if ok-to-stop says the PGs first need to settle, but that's fine.

Reply:
I think that should be possible, but I am concerned about corner cases. I'd like to ask @leseb and @liewegas if adding new OSDs to a Ceph cluster while some OSDs are being stopped/restarted would risk any corner cases where availability might suffer.